neon/polyfills.js

```diff
@@ -0,0 +1,248 @@
+        } else {
+          throw Error(
+            "createHash only supports sha256 or md5 in this environment, not ${type}.",
+          );
+        }
+        if (encoding === "hex") {
+          return prom.then((arrayBuf) => Buffer.from(arrayBuf).toString("hex"));
+        } else if (encoding) {
+          throw Error(
+            `createHash only supports hex encoding or unencoded in this environment, not ${encoding}`,
+          );
+        } else {
+          return prom;
+        }
+      },
+    }),
+  }),
+  createHmac: (type, key) => ({
+    update: (x) => ({
+      digest: async () =>
+        Buffer.from(
+          await Crypto.subtle.sign(
+            "HMAC",
+            await Crypto.subtle.importKey(
+              "raw",
+              key,
+              { name: "HMAC", hash: "SHA-256" },
+              false,
+              ["sign"],
+            ),
+            textEncoder.encode(x),
+          ),
+        ),
+    }),
+  }),
+};
+
+export const performance = globalThis.performance;
+
+export const process = {
+  env: {},
+};
+
+export const os = {
+  userInfo() {
+    return { username: "postgres" };
+  },
+};
+
+export const fs = {
+  readFile() {
+    throw new Error("Reading files not supported on CloudFlare");
+  },
+};
+
+export const net = {
+  isIP: (x) => (IPv4Reg.test(x) ? 4 : IPv6Reg.test(x) ? 6 : 0),
+  Socket,
+};
+
+export { setImmediate, clearImmediate };
+
+export const tls = {
+  connect({ socket, servername }) {
+    socket.startTls(servername);
+    return socket;
+  },
+};
+
+const wsCache = new Map();
+
+function Socket() {
+  const tcp = Object.assign(new EventEmitter(), {
+    readyState: "open",
+    ws: null,
+    writeBuffer: undefined,
+    connect,
+    startTls,
+    write,
+    end,
+    destroy,
+    read,
+  });
+
+  return tcp;
+
+  async function connect(port, host) {
+    const handleWebSocketOpen = () => {
+      tcp.pending = false;
+      tcp.emit("connect");
+      tcp.emit("ready");
+    };
+    const configureWebSocket = (ws, immediateOpen = false) => {
+      ws.binaryType = "arraybuffer";
+
+      ws.addEventListener("error", (err) => {
+        tcp.emit("error", err);
+        tcp.emit("close");
+      });
+
+      ws.addEventListener("message", (msg) => {
+        const buffer = Buffer.from(msg.data);
+        tcp.emit("data", buffer);
+      });
+
+      ws.addEventListener("close", () => {
+        tcp.emit("close");
+      });
+
+      if (immediateOpen) handleWebSocketOpen();
+      else ws.addEventListener("open", handleWebSocketOpen);
+    };
+    try {
+      tcp.readyState = "opening";
+      const rootURL = host + "/v2" + "?address=" + host + ":" + port;
+      const socketURL = "wss://" + rootURL;
+      tcp.ws = new WebSocket(socketURL);
+      configureWebSocket(tcp.ws);
+    } catch (err) {
+      console.log("err", err);
+      error(err);
+    }
+  }
+
+  async function startTls(host) {
+    throw new Error("Postgres SSL connections are not supported yet");
```

Comment:
The connection goes over secure web sockets (`wss://`) so we don't need to double encrypt the content. As such, we don't support tls from postgres.